### PR TITLE
[UNDERTOW-1795] Always sort cookies

### PR DESCRIPTION
**Pull request by ropalka**
https://issues.redhat.com/browse/UNDERTOW-1795

### CODE REVIEW

**Review by scottmarlow — COMMENTED** [review]
:+1:

**Review by fl4via — APPROVED** [review]

File: core/src/main/java/io/undertow/server/handlers/CookieImpl.java
Comment (bstansberry) [inline core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: @ropalka @fl4via 

Should this be the default impl in the interface? AFAICT it's entirely using calls to interface methods, so it could be a default impl.  Doing it that way avoids making an incompatible change in a public interface.

Comment (carterkozak) [reply core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: From the peanut gallery:

Thoughts on using a constant `Comparator<Cookie>` to provide to the TreeSet to avoid issues with incorrect custom cookie implementations?

Comment (bstansberry) [reply core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: Note that even though I commented here but don't block waiting for a PR approval from me.

Comment (bstansberry) [reply core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: @carterkozak I'll defer to @ropalka on that one. Seems reasonable.

Comment (ropalka) [reply core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: That was my initial implementation @bstansberry but I didn't like
default methods cannot be final. So I moved it to cookie impl classes instead.

But thinking about it again it introduced good-for-nothing code duplicities so
I reworked it back again as you propose. Additionally we want to forbid
subclasses to break the contract (same applies to equals() & hashCode() methods)
so we override comparteTo() method in subclasses to delegate to default method impl.

Comment (ropalka) [reply core/src/main/java/io/undertow/server/handlers/CookieImpl.java:207]: Thanks @carterkozak for your proposal. That is also alternative.
I prefer compareTo() to be co-located in classes where
equals(), hashCode() & toString() methods are declared.

**Review by fl4via — APPROVED** [review]

**Pull request merged**
